### ISSUE TITLE
Pod in Failed state, but Nextflow still waiting

### ISSUE THREAD

**Issue by mes5k** [nextflow-io/nextflow#2428]
## Bug report 

Using the Kubernetes executor we've produced a situation where a process _pod_ will fail, but nextflow will continue waiting on the pod forever.

### Expected behavior and actual behavior

We inject an initContainer into our nextflow process pods for internal purposes. Everything works as expected when the initContainer succeeds (i.e. the normal nextflow process pod starts and executes as expected), but when the initContainer fails, nextflow keeps waiting for the pod even though the pod phase is "Failed". 

### Steps to reproduce the problem

Use special magic to add an initContainer to process pods and then have the initContainer fail.

### Program output 

I've attached an example process pod YAML below.

In that YAML can see that `containerStatuses` has 
```
    state:
      waiting:
        reason: PodInitializing
```
which triggers [this line in K8sClient.groovy](https://github.com/nextflow-io/nextflow/blob/master/modules/nextflow/src/main/groovy/nextflow/k8s/client/K8sClient.groovy#L256), which causes nextflow to continue waiting.  However, you can also see that the overall pod phase is "Failed".
```
  phase: Failed
```

I think this can be fixed by adding an additional check for "Failed" to [checkInvalidWaitingState](https://github.com/nextflow-io/nextflow/blob/master/modules/nextflow/src/main/groovy/nextflow/k8s/client/K8sClient.groovy#L300).  I'll put together a patch.

### Environment 

* Nextflow version: 21.04.1

### Additional context
[nextflow-not-failed.yaml.txt](https://github.com/nextflow-io/nextflow/files/7471696/nextflow-not-failed.yaml.txt)

**Comment by mes5k** [comment]
@pditommaso Unfortunately, there still appears to be a problem here. The problem happens when `errorStrategy` is set to `ignore`. In this case if an init container fails Nextflow will still hang.  

For a normal process failure, things work as I'd expect. If my process runs `exit 1` the process fails, the pod goes into an error state, and Nextflow closes all channels, completes, and cleans up the error pod.  

On the other hand, if I inject an init container that runs `exit 1`, then Nextflow correctly identifies the failed pod (with my original patch code) and logs that the process is being ignored, but somehow that task is never cleaned up, which means the output channel for that process remains open, which causes Nextflow to hang.  My guess is that the exception produced by the failed init container comes too early in the task processing for Nextflow to properly clean up, but I'm not really sure.

I'm not quite sure where I should be looking to make sure that the task gets properly cleaned up.  If you can provide any pointers or hints on where to look I can try to produce another patch.  Thanks for any guidance!

**Comment by pditommaso** [comment]
I suspect the `initContainer` has a different (API) behaviour that should be managed properly. 

Leave this in your hands for further investigation

**Comment by mes5k** [comment]
We just found another way that can trigger this problem. We ran into this error: https://github.com/containerd/containerd/issues/4604 for a Nextflow process pod. The pod in question appeared in state "CreateContainerError". An exception was thrown which Nextflow logged and ignored (because of errorStrategy ignore), but the task itself never got cleaned up so Nextflow main was hanging as above.

The good news is that I've finally found the time to dig in and I think I've got a fix. A PR for your consideration should be forthcoming.
